### PR TITLE
Add support for custom name in Spring stereotype annotations

### LINKED ISSUE #1427: @Mapper(componentModel = "spring") doesn't support custom name for Spring Service Annotation
Can't specify a name for the Spring Service annotation
If the interface definition something like

    @Mapper(componentModel = "spring", componentModelName = "cakeMapperV2")
     public interface CakeMapper{
         Cake map(Cupcake cupcake);
     }

     @Service("cakeMapperV2")
     public class CakeMapperImpl {
     }

### PR DESCRIPTION
**Pull request by jccampanero**
Adds support for defining a custom name for the Spring stereotype annotations (`@Component`, `@Service`) in the generated `Mapper` when using the Spring component model. 

Fixes #1427

### CODE REVIEW

**Comment by jccampanero** [comment]
Thank you very much for the feedback @Zegveld.

I absolutely agree with you: I initially thought of including checks for `@Repository` and `@Controller`, but they seem like a less obvious use case to me. But you are completely right, the way to go should be traversing the different annotation tree looking for the occurrence of `@Component` within it, directly or through the annotation hierarchy. I will review `RepeatableAnnotations` as you suggested looking for reference.

Again, thank you very much for the feedback, and sorry for this first baby step.

**Comment by jccampanero** [comment]
@Zegveld Sorry for the late reply.

I have pushed some new changes.

Per your suggestion, I'm trying to process the `Mapper` instance by checking for the existence of the `@Component` annotation through the entire annotation hierarchy.

Please let me know what additional work needs to be done.

Thank you very much.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/decorator/SpringAnnotateWithDecoratorMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        CustomerSpringComponentQualifiedMapper.class,
+        UpperCaseCustomerNameSpringDefaultMapperDecorator.class,
+        CustomerSpringDefaultMapper.class,
+        GenderSpringDefaultMapper.class
+})
+@IssueKey( "1427" )
+@ComponentScan(basePackageClasses = CustomerSpringDefaultMapper.class)
+@Configuration
+@WithSpring
+public class SpringAnnotateWithDecoratorMapperTest {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/decorator/SpringAnnotateWithDecoratorMapperTest.java:47]: Do we really need this test? I think that the most important parts are to verify that the right things are in the generated code.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    @Autowired
+    @Qualifier( "AnnotateWithCustomStereotype" )
+    private CustomerSpringDefaultMapper annotateWithCustomStereotypeCustomerMapper;
+
+    @RegisterExtension
+    final GeneratedSource generatedSource = new GeneratedSource();
+    private ConfigurableApplicationContext context;
+
+    @BeforeEach
+    public void springUp() {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:79]: I don't really think that we need to test this in the spring context. We can just validate the generated sources.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    @ProcessorTest
+    public void shouldHaveControllerAnnotatedQualifiedMapper() {
+
+        // then
+        assertThat( annotateWithControllerCustomerMapper ).isNotNull();
+        generatedSource.forMapper( CustomerSpringControllerQualifiedMapper.class )
+                .content()
+                .contains( "@Controller(value = \"AnnotateWithController\")" )
+                .doesNotContain( "@Component" + System.lineSeparator() );
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:111]: We don't need the line separator here.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    @ProcessorTest
+    public void shouldHaveServiceAnnotatedQualifiedMapper() {
+
+        // then
+        assertThat( annotateWithServiceCustomerMapper ).isNotNull();
+        generatedSource.forMapper( CustomerSpringServiceQualifiedMapper.class )
+                .content()
+                .contains( "@Service(value = \"AnnotateWithService\")" )
+                .doesNotContain( "@Component" + System.lineSeparator() );
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:123]: We don't need the line separator here.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    @ProcessorTest
+    public void shouldHaveRepositoryAnnotatedQualifiedMapper() {
+
+        // then
+        assertThat( annotateWithRepositoryCustomerMapper ).isNotNull();
+        generatedSource.forMapper( CustomerSpringRepositoryQualifiedMapper.class )
+                .content()
+                .contains( "@Repository(value = \"AnnotateWithRepository\")" )
+                .doesNotContain( "@Component" + System.lineSeparator() );
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:135]: We don't need the line separator here.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    @ProcessorTest
+    public void shouldHaveCustomStereotypeAnnotatedQualifiedMapper() {
+
+        // then
+        assertThat( annotateWithCustomStereotypeCustomerMapper ).isNotNull();
+        generatedSource.forMapper( CustomerSpringCustomStereotypeQualifiedMapper.class )
+                .content()
+                .contains( "@CustomStereotype(value = \"AnnotateWithCustomStereotype\")" )
+                .doesNotContain( "@Component" + System.lineSeparator() );
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:147]: We don't need the line separator here.

**Review by filiphr — CHANGES_REQUESTED** [review]
The implementation looks good @jccampanero. I only have some points regarding the tests.

I do not think that we should complicate them that much. Do we really need to test this entirely using this in the Spring context. I think that it is enough to what we are doing in `SpringAnnotateWithMapperTest` without the spring context testing. We could also simplify the mappers by not doing `Mapper#uses` and without extending from other mappers. We can have empty mappers for the tests.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java (hunk 3 above)
Comment (jccampanero) [reply processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:111]: I totally agree with you, thank you.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java (hunk 2 above)
Comment (jccampanero) [reply processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/qualified/SpringAnnotateWithMapperTest.java:79]: Thank you for point that out @filiphr. I updated the code to test only the generated source.

**Comment by jccampanero** [comment]
Thank you very much for the code review @filiphr, I appreciate it a lot.

I updated the code with the changes you requested.

Please let me know if any additional changes need to be made.

File: processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/decorator/SpringAnnotateWithDecoratorMapperTest.java (hunk 1 above)
Comment (jccampanero) [reply processor/src/test/java/org/mapstruct/ap/test/injectionstrategy/spring/annotateWith/decorator/SpringAnnotateWithDecoratorMapperTest.java]: I removed the code and refactor `SpringAnnotateWithMapperTest` and related stuff to its original package. I will include ir t again if necessary.

Thank you very much for the feedback @filiphr.

**Comment by filiphr** [comment]
Thanks a lot for your patience @jccampanero. We've integrated this into main.

**Comment by jccampanero** [comment]
Thank you very much @filiphr. Please, on the contrary, thank you very much for your time and for allowing me to contribute to this wonderful library.

**Pull request merged**